libtsuba/src/tsuba.cpp

```diff
@@ -307,6 +308,88 @@ katana::CopyRDG(
   return katana::ResultSuccess();
 }
 
+katana::Result<void>
+katana::WriteRDGPartHeader(
+    std::vector<katana::RDGPropInfo> node_properties,
+    std::vector<katana::RDGPropInfo> edge_properties,
+    katana::EntityTypeManager& node_entity_type_manager,
+    katana::EntityTypeManager& edge_entity_type_manager,
+    const std::string& node_entity_type_id_array_path,
+    const std::string& edge_entity_type_id_array_path, uint64_t num_nodes,
+    uint64_t num_edges, const std::string& topology_path,
+    const std::string& rdg_dir) {
+  auto manifest = RDGManifest();
+  // manifest = manifest.NextVersion(1, 0, false, RDGLineage());
+  katana::Uri rdg_dir_uri = KATANA_CHECKED(katana::Uri::Make(rdg_dir));
+  manifest.set_dir(rdg_dir_uri);
+  manifest.set_viewtype(katana::kDefaultRDGViewType);
+  auto part_header = katana::RDGPartHeader();
+
+  // Setup partition metadata, for now assume only one partition
+  katana::PartitionMetadata part_meta;
+  part_meta.num_global_nodes_ = num_nodes;
+  part_meta.max_global_node_id_ = num_nodes - 1;
+  part_meta.num_global_edges_ = num_edges;
+  part_meta.num_edges_ = num_edges;
+  part_meta.num_nodes_ = num_nodes;
+  part_meta.num_owned_ = num_nodes;
+  part_meta.is_outgoing_edge_cut_ = true;
+
+  // Create vector that is needed by part_header for prop_info, do this for both node and edges
+  std::vector<katana::PropStorageInfo> node_props;
+  node_props.reserve(node_properties.size());
+  for (auto rdg_prop_info : node_properties) {
+    node_props.emplace_back(katana::PropStorageInfo(
+        rdg_prop_info.property_name, rdg_prop_info.property_path));
+  }
+
+  std::vector<katana::PropStorageInfo> edge_props;
+  edge_props.reserve(edge_properties.size());
+  for (auto rdg_prop_info : edge_properties) {
+    edge_props.emplace_back(katana::PropStorageInfo(
+        rdg_prop_info.property_name, rdg_prop_info.property_path));
+  }
+
+  // Set the node and edge prop info lists
+  part_header.set_node_prop_info_list(std::move(node_props));
+  part_header.set_edge_prop_info_list(std::move(edge_props));
+
+  // Set the entity type managers for nodes and edges
+  part_header.StoreNodeEntityTypeManager(node_entity_type_manager);
+  part_header.StoreEdgeEntityTypeManager(edge_entity_type_manager);
+
+  // Set the paths for the entity type id arrays
+  part_header.set_node_entity_type_id_array_path(
+      node_entity_type_id_array_path);
+  part_header.set_edge_entity_type_id_array_path(
+      edge_entity_type_id_array_path);
+
+  // Set the topology metadata
+  part_header.MakePartitionTopologyMetadataEntry(topology_path);
+
+  katana::RDGTopologyManager topo_manager = KATANA_CHECKED(
+      katana::RDGTopologyManager::Make(part_header.topology_metadata()));
+
+  // get the metadata we need from the topology file
+  KATANA_CHECKED(topo_manager.ExtractMetadata(
```

Comment:
Before this is called, we have written the topology ourselves right? So we shouldn't have to extract the metadata from the topology, since we have the data we need in memory already. 

Since all we care about at this point is the part header, we actually don't need a topology manager at all. 

Instead, we should just fill the PartitionTopologyMetadataEntry ourselves:
```

  // Set the topology metadata
  PartitioTopologyMetadataEntry * entry =  part_header.MakePartitionTopologyMetadataEntry(topology_path);

  // update our metadata entry with what we loaded
  // Assumes the topology is a standard csr topology
  topology_state_ = TopologyKind::kCSR;
  transpose_state_ = TransposeKind::kNo;
  edge_sort_state_ = EdgeSortKind::kAny;
  node_sort_state_ = NodeSortKind::kAny;

 entry->Update(
      num_edges_, num_nodes_,
      /*edge_index_to_property_index_map_present=*/false,
      /*node_index_to_property_index_map_present=*/false,
      /*edge_condensed_type_id_map_size_=*/0,
      /*edge_condensed_type_id_map_present=*/false,
      /*node_condensed_type_id_map_size_=*/0,
      /*node_condensed_type_id_map_present=*/false, 
      topology_state_,
      transpose_state_, 
      edge_sort_state_, 
      node_sort_state_);
```

You could abstract this out into a `FillCSRMetadataEntry` so both this code and `MapMetadataExtract` can use it, if you'd like.

Reply:
A little confused here - could you clarify. So yes, at this point, we have written the topology out to disk, however I don't think that data is in memory already? I figured that I would use this method with the Topology manager in case we do anything else in the future to add on metadata with.

Reply:
Thanks for the clarification. I think that makes sense. I'lll refactor out the topology metadata setting in another method.